finality-aleph/src/sync/forest/mod.rs

```diff
@@ -289,52 +306,107 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
             self.prune(&id);
         }
         self.compost_bin.retain(|k| k.number() > level);
+        self.justified_blocks.retain(|k, _| k > &level);
     }
 
     /// Attempt to finalize one block, returns the correct justification if successful.
-    pub fn try_finalize(&mut self) -> Option<J> {
-        for child_id in self.root_children.clone().into_iter() {
-            if let Some(VertexWithChildren { vertex, children }) = self.vertices.remove(&child_id) {
+    pub fn try_finalize(&mut self, number: &u32) -> Option<J> {
+        // cached as ready
+        if let Some(id) = self.justified_blocks.get(number) {
+            if let Some(VertexWithChildren { vertex, children }) = self.vertices.remove(id) {
                 match vertex.ready() {
+                    // ready indeed
                     Ok(justification) => {
-                        self.root_id = child_id;
+                        self.root_id = id.clone();
                         self.root_children = children;
                         self.prune_level(self.root_id.number());
                         return Some(justification);
                     }
+                    // cache corrupted, we'll wait for it to fix itself
                     Err(vertex) => {
                         self.vertices
-                            .insert(child_id, VertexWithChildren { vertex, children });
+                            .insert(id.clone(), VertexWithChildren { vertex, children });
                     }
                 }
             }
         }
         None
     }
 
-    /// How much interest we have for the block.
-    pub fn state(&mut self, id: &BlockIdFor<J>) -> Interest<I> {
+    /// Returns the BranchKnowledge regarding the given block id,
+    /// or None if any errors are encountered.
+    fn branch_knowledge(&mut self, mut id: BlockIdFor<J>) -> Option<BranchKnowledge<J>> {
+        use VertexHandle::*;
+        let mut maybe_parent_id;
+        // traverse ancestors till we reach something imported or a parentless vertex
+        // avoid infinite loop by limiting the number of iterations to max possible value
+        for _ in 0..MAX_DEPTH {
```

Comment:
That's really not better. :/

Reply:
It's even worse! Will change to `loop`, as discussed.